src/lib/Inline.hs

```diff
@@ -0,0 +1,475 @@
+    --   effectful bindings that are already in the substitution.
+    resolveWorkConservation (OccInfoImpure _) _ = NoInline
+    -- This function for detecting available indexing depth reports all decls as
+    -- aborting indexing, even if the decl will be inlined away later.
+    --
+    -- For example, the binding
+    --   xs = for i.
+    --     ys = for j. <something>
+    --     for k. ys.k
+    -- really does have two available indexing levels, because the ys binding is
+    -- inlinable into the result of the outer `for`; but the below function will
+    -- not detect it, and thus we will decline to inline `xs` into a context
+    -- where two levels of indexing are required.  However, we can recover the
+    -- desired effect by running a second pass of occurrence analysis and
+    -- inlining.
+    --
+    -- A more subtle case occurs with
+    --   xs = for i.
+    --     ys = for j. <something>
+    --     view k. ys.k
+    -- (note the `view`).  In this case, if we went ahead and inlined `xs` into
+    -- a context that required two levels of indexing, we would temporarily
+    -- duplicate work, but a second pass of occurence analysis and inlining
+    -- would fix it by inlining `ys`.  However, if we decline to inline here,
+    -- running inlining on the body of `for i.` will not inline into the `view`,
+    -- because in general views are supposed to be duplicable all over.  Maybe
+    -- the solution is to just not have view atoms in the IR by this point,
+    -- since their main purpose is to force inlining in the simplifier, and if
+    -- one just stuck like this it has become equivalent to a `for` anyway.
+    ixDepthExpr :: Expr r n -> Int
+    ixDepthExpr (Hof (For _ _ (Lam (LamExpr _ body)))) = 1 + ixDepthBlock body
+    ixDepthExpr (Atom (TabLam (TabLamExpr _ body))) = 1 + ixDepthBlock body
+    ixDepthExpr _ = 0
+    ixDepthBlock :: Block r n -> Int
+    ixDepthBlock (exprBlock -> (Just expr)) = ixDepthExpr expr
+    ixDepthBlock (AtomicBlock result) = ixDepthExpr $ Atom result
+    ixDepthBlock _ = 0
+
+-- Should we decide to inline this binding wherever it appears, before we even
+-- know the expression?  "Yes" only if we know it only occurs once, and in a
+-- context where inlining it does not duplicate work.
+preInlineUnconditionally :: LetAnn -> Bool
+preInlineUnconditionally = \case
+  PlainLet -> False  -- "Missing occurrence annotation"
+  NoInlineLet -> False
+  OccInfoPure (UsageInfo s (0, d)) | s <= One && d <= One -> True
+  OccInfoPure _ -> False
+  OccInfoImpure _ -> False
+
+data Context (from::E) (to::E) (o::S) where
+  Stop :: Context e e o
+  TabAppCtx :: NonEmpty (SAtom i) -> Subst InlineSubstVal i o
+            -> Context SExpr e o -> Context SExpr e o
+  EmitToAtomCtx :: Context SAtom e o -> Context SExpr e o
+  EmitToNameCtx :: Context SAtomName e o -> Context SAtom e o
+  RepWrap :: GenericE e1 => Context e1 e2 o -> Context (RepE e1) e2 o
```

Comment:
It would be helpful to add some comments here, because I don't fully understand what this does. I feel like you've defunctionalized something, but I don't know what that is.

Reply:
Made an attempt; PTAL.